### PR TITLE
test(integration-em) add basic election manager integration tests

### PR DESCRIPTION
**Pull request by carolinemodic**
Adds two basic election manager tests. In the bsd tests both bsd and module-scan seemed to get killed properly after the completion of the tests but here it doesn't seem like module-smartcards or module-convertor-sems are ever getting killed... I think because they are python servers? @eventualbuddha lmk if you have any ideas on how to fix this. 

Configuration - Loads election manager by converting sems files

https://user-images.githubusercontent.com/14897017/123462872-b4840780-d59f-11eb-9572-b497c3f871d1.mp4

Tallies - Loads election manager, imports CVR file, converts to SEMS and verifies the output 

https://user-images.githubusercontent.com/14897017/123462887-b77ef800-d59f-11eb-9ad3-222f179ab39e.mp4

### CODE REVIEW

File: libs/fixtures/src/data/electionMultiPartyPrimary/semsFiles/standard.csv.ts [hunk 1]
```diff
@@ -1,364 +1,364 @@
-export default `"sample-county","precinct-1","assistant-mayor-contest-liberty","Assistant Mayor","0","Li","0","Write-in","0","Li","30",
-"sample-county","precinct-1","assistant-mayor-contest-liberty","Assistant Mayor","0","Li","1","Times Over Voted","0","Li","30",
-"sample-county","precinct-1","assistant-mayor-contest-liberty","Assistant Mayor","0","Li","2","Times Under Voted","0","Li","30",
+export default `"sample-county","precinct-1","assistant-mayor-contest-liberty","Assistant Mayor","0","Li","0","Write-in","0","NP","30",
```
Comment (carolinemodic) [inline libs/fixtures/src/data/electionMultiPartyPrimary/semsFiles/standard.csv.ts:1]: In the past we had a bug where when an election had a party with the id of "0" it got mislabelled as the party in write-in/undervotes/overvotes. This was fixed already in module-converter-sems but the SEMS fixtures files that were autogenerated were never updated after this bug fix.

File: .circleci/config.yml [hunk 2]
```diff
@@ -355,6 +355,25 @@ jobs:
           command: |
             pnpm --dir integration-testing/bsd test
 
+  test-integration-testing-election-manager:
+    docker:
+      - image: circleci/python:3.7.7-node-browsers
```
Comment (eventualbuddha) [inline .circleci/config.yml:360]: What version of node is this using?

File: integration-testing/election-manager/.stylelintrc-css.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// Config for stylelint parsing CSS
```
Comment (eventualbuddha) [inline integration-testing/election-manager/.stylelintrc-css.js:1]: Do we need this file?

File: integration-testing/election-manager/.stylelintrc.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// Ensure we use all available babel parser plugins, but not the type annotation
```
Comment (eventualbuddha) [inline integration-testing/election-manager/.stylelintrc.js:1]: Do we need this file?

File: integration-testing/election-manager/Makefile [hunk 5]
```diff
@@ -0,0 +1,31 @@
+ELECTION_MANAGER := ../../apps/election-manager
+SMARTCARDS := ../../apps/module-smartcards
+SEMS := ../../apps/module-converter-ms-sems
+export PIPENV_VENV_IN_PROJECT=1
+export MODULE_SCAN_WORKSPACE=/tmp
+export MOCK_SCANNER_FILES=,,
```
Comment (eventualbuddha) [inline integration-testing/election-manager/Makefile:6]: Not relevant?

File: integration-testing/election-manager/Makefile [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+clean-downloads:
+	rm cypress/downloads/*
+
+build-election-manager:
+	make -C $(ELECTION_MANAGER) build; \
```
Comment (eventualbuddha) [inline integration-testing/election-manager/Makefile:12]: Unnecessary continuation? Or keeping it to make it easier to add another line?

```suggestion
	make -C $(ELECTION_MANAGER) build
```

File: integration-testing/election-manager/cypress/integration/configuration.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    cy.contains('Convert from SEMS files').click()
+    cy.get('input[name="SEMS main file"]').attachFile('semsMainFile.txt')
+    cy.get('input[name="SEMS candidate mapping file"]').attachFile(
+      'semsCandidateMappingFile.txt'
+    )
+    cy.contains('Election loading')
+    cy.contains('Special Election for Senate 15')
+    cy.contains('Election Hash: f30480cc99')
+    // eslint-disable-next-line cypress/no-unnecessary-waiting
+    cy.wait(500)
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/integration/configuration.ts:13]: Why?

File: integration-testing/election-manager/cypress/integration/tallies.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import { electionMultiPartyPrimaryWithDataFiles } from '@votingworks/fixtures'
+
+describe('Election Manager can create SEMS tallies', () => {
+  it('Election Manager can tally results properly', () => {
+    cy.visit('/')
+    cy.get('input[type="file"]').attachFile(
+      'electionMultiPartyPrimarySample.json'
+    )
+    cy.contains('Election loading')
+    // cy.contains('Election Hash: abdfbe6a58')
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/integration/tallies.ts:10]: Uncomment or delete?

File: integration-testing/election-manager/cypress/integration/tallies.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    cy.get('input[data-testid="manual-input"]').attachFile(
+      'multiPartyPrimaryCVRResults.jsonl'
+    )
+    cy.get('[data-testid="total-ballot-count"]').within(() =>
+      cy.contains('4,530')
+    )
+    cy.contains('Save Results File').click()
+    cy.get('[data-testid="manual-export"]').click()
+    cy.contains('Results Saved')
+    cy.task('readMostRecentFile', 'cypress/downloads').then((fileContent) => {
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/integration/tallies.ts:22]: Does cypress recommend not using async/await?

File: integration-testing/election-manager/cypress/integration/tallies.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+      const recievedLines = (fileContent as string).split('\r\n')
+      let i = 0
+      for (const expectedLine of electionMultiPartyPrimaryWithDataFiles.semsData.split(
+        '\n'
+      )) {
+        expect(recievedLines[i].replace('\n', ': ')).to.eqls(expectedLine)
+        i += 1
+      }
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/integration/tallies.ts:30]:
```suggestion
      for (const [i, expectedLine] of electionMultiPartyPrimaryWithDataFiles.semsData.split(
        '\n'
      ).entries()) {
        expect(recievedLines[i].replace('\n', ': ')).to.eqls(expectedLine)
      }
```

File: integration-testing/election-manager/cypress/plugins/index.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * @type {Cypress.PluginConfig}
+ */
+// eslint-disable-next-line no-unused-vars
+module.exports = (on, config) => {
+  // `on` is used to hook into various events Cypress emits
+  // `config` is the resolved Cypress config
+  on('task', {
+    async readMostRecentFile(directoryPath) {
+      const files = fs.readdirSync(directoryPath)
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/plugins/index.js:27]: If we're using `async` already, make this `await`? Also requires changing the import to `import { promises as fs } from 'fs'`.

```suggestion
      const files = await fs.readdir(directoryPath)
```

File: integration-testing/election-manager/cypress/plugins/index.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      files.sort((fileOne, fileTwo) => {
+        const { ctime: ctimeOne } = fs.statSync(
+          path.join(directoryPath, fileOne)
+        )
+        const { ctime: ctimeTwo } = fs.statSync(
+          path.join(directoryPath, fileTwo)
+        )
+        return ctimeTwo - ctimeOne
+      })
+      return fs.readFileSync(path.join(directoryPath, files[0]), 'utf-8')
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/plugins/index.js:38]: Ditto `await`.

File: integration-testing/election-manager/cypress/plugins/index.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+ */
+// eslint-disable-next-line no-unused-vars
+module.exports = (on, config) => {
+  // `on` is used to hook into various events Cypress emits
+  // `config` is the resolved Cypress config
+  on('task', {
+    async readMostRecentFile(directoryPath) {
+      const files = fs.readdirSync(directoryPath)
+      // Sort files by ctime to put the most recent first
+      files.sort((fileOne, fileTwo) => {
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/plugins/index.js:29]: Using sort like this mutates `files` and is something we generally try to avoid (in fact, we have a custom eslint rule in our apps/libs prohibiting this). You could do `const mostRecentFile = [...files].sort(…)[0]`, but since we only want the most recent you could do this (assumes `await` changes made elsewhere):

```ts
const paths = files.map((file) => path.join(directoryPath, file))
const ctimes = await Promise.all(paths.map(async (p) => (await fs.stat(p)).ctime)
const mostRecentCtime = Math.max(ctimes)
const mostRecentPath = paths[ctimes.indexOf(mostRecentCtime)]
return await fs.readFile(mostRecentPath, 'utf-8')
```

File: integration-testing/election-manager/cypress/support/commands.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+//
+//
+// -- This will overwrite an existing command --
+// Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+import 'cypress-file-upload'
+
+Cypress.Commands.add('readMostRecentFile', (directoryPath) => {
+  console.log('HELLO')
+  return ''
+})
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/support/commands.js:31]: Does this do anything?

File: integration-testing/election-manager/cypress/tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    "noEmit": true,
+    "noFallthroughCasesInSwitch": true,
+    "resolveJsonModule": true,
+    "skipLibCheck": true,
+    "strict": true,
+    "target": "es2015"
+  },
+  "include": [
+    "../node_modules/cypress",
+    "../node_modules/@testing-library/cypress",
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/tsconfig.json:23]: Is this intentional? Are there `.ts` files in these locations?

File: integration-testing/election-manager/package.json [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    "eslint-config-prettier": "^6.10.0",
+    "eslint-import-resolver-node": "^0.3.4",
+    "eslint-plugin-cypress": "^2.11.2",
+    "eslint-plugin-flowtype": "^5.2.0",
+    "eslint-plugin-import": "^2.20.1",
+    "eslint-plugin-jest": "^23.6.0",
+    "eslint-plugin-jsx-a11y": "^6.2.3",
+    "eslint-plugin-prettier": "^3.1.2",
+    "eslint-plugin-vx": "workspace:*",
+    "husky": "^4.2.1",
```
Comment (eventualbuddha) [inline integration-testing/election-manager/package.json:46]: Unused.
```suggestion
```

File: integration-testing/election-manager/package.json [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    "is-ci-cli": "^2.1.2",
+    "jest-environment-jsdom-sixteen": "^1.0.3",
+    "lint-staged": "^10.0.7",
```
Comment (eventualbuddha) [inline integration-testing/election-manager/package.json:49]: Unused.
```suggestion
```

File: integration-testing/election-manager/package.json [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+    "eslint-plugin-cypress": "^2.11.2",
+    "eslint-plugin-flowtype": "^5.2.0",
+    "eslint-plugin-import": "^2.20.1",
+    "eslint-plugin-jest": "^23.6.0",
+    "eslint-plugin-jsx-a11y": "^6.2.3",
+    "eslint-plugin-prettier": "^3.1.2",
+    "eslint-plugin-vx": "workspace:*",
+    "husky": "^4.2.1",
+    "is-ci-cli": "^2.1.2",
+    "jest-environment-jsdom-sixteen": "^1.0.3",
```
Comment (eventualbuddha) [inline integration-testing/election-manager/package.json:48]: Unused?
```suggestion
```

File: .circleci/config.yml (hunk 2 above)
Comment (eventualbuddha) [reply .circleci/config.yml:360]: Answering my own question: 12.18.1. This should be fine, but we should be careful when we update node versions or change this image that this may not be what we expect.

**Review by eventualbuddha — APPROVED** [review]
There's some cleanup to do here, but like the bsd one this looks fine overall. I do worry a little bit about the fixtures, specifically not really knowing how they got built and how to change them. Documentation for that would be great.

**Comment by eventualbuddha** [comment]
This fails for me locally running with `pnpm test:ci`:

```
  1) Election Manager can create SEMS tallies
       Election Manager can tally results properly:

      AssertionError: expected '"sample-county","precinct-1","assistant-mayor-contest-liberty","Assistant Mayor","0","Li","0","Write-in","0","NP","30",' to deeply equal '"sample-county","precinct-1","assistant-mayor-contest-liberty","Assistant Mayor","0","Li","0","Write-in","0","Li","30",'
      + expected - actual

      -'"sample-county","precinct-1","assistant-mayor-contest-liberty","Assistant Mayor","0","Li","0","Write-in","0","NP","30",'
      +'"sample-county","precinct-1","assistant-mayor-contest-liberty","Assistant Mayor","0","Li","0","Write-in","0","Li","30",'
      
      at Context.eval (http://localhost:3000/__cypress/tests?p=cypress/integration/tallies.ts:24778:65)
```

File: integration-testing/election-manager/Makefile [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+build-converter-ms-sems:
+	sudo apt update -y; \
```
Comment (eventualbuddha) [inline integration-testing/election-manager/Makefile:15]: Why is this here?

File: integration-testing/election-manager/Makefile [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+	make -C $(SEMS) install-dependencies; \
+	make -C $(SEMS) install-dev-dependencies; \
+	make -C $(SEMS) build; \
+
+build-smartcards:
+	sudo apt update -y; \
```
Comment (eventualbuddha) [inline integration-testing/election-manager/Makefile:21]: Why is this here?

File: .circleci/config.yml (hunk 2 above)
Comment (carolinemodic) [reply .circleci/config.yml:360]: Ah yeah, when I copied the circle setup from other node projects running the python apps failed so I just used this as a way to get both node and python in the image

File: integration-testing/election-manager/cypress/integration/tallies.ts (hunk 8 above)
Comment (carolinemodic) [reply integration-testing/election-manager/cypress/integration/tallies.ts:10]: Woops yeah I changed elections and forgot to update this.

File: integration-testing/election-manager/cypress/integration/tallies.ts (hunk 9 above)
Comment (carolinemodic) [reply integration-testing/election-manager/cypress/integration/tallies.ts:22]: yeah I believe you can't use async/await as cypress has its own similar task queue. https://docs.cypress.io/guides/core-concepts/introduction-to-cypress#Commands-Are-Asynchronous

File: integration-testing/election-manager/cypress/support/commands.js (hunk 14 above)
Comment (carolinemodic) [reply integration-testing/election-manager/cypress/support/commands.js:31]: lol sorry I should have self-reviewed this before sending out. I could not figure out how to trigger a command here, will remove

File: integration-testing/election-manager/cypress/tsconfig.json (hunk 15 above)
Comment (carolinemodic) [reply integration-testing/election-manager/cypress/tsconfig.json:23]: This is copied from bmd, this tsconfig is inside the cypress/ subdirectory so these are valid paths.

File: integration-testing/election-manager/cypress/plugins/index.js [hunk 21]
```diff
@@ -27,9 +27,9 @@ module.exports = (on, config) => {
       const files = await fs.readdir(directoryPath)
       const paths = files.map((file) => path.join(directoryPath, file))
       const ctimes = await Promise.all(
-        paths.map(async (p) => (await fs.stat(p)).ctime)
+        paths.map(async (p) => (await fs.stat(p)).ctime.getTime())
       )
-      const mostRecentCtime = Math.max(ctimes)
+      const mostRecentCtime = Math.max(...ctimes)
```
Comment (eventualbuddha) [inline integration-testing/election-manager/cypress/plugins/index.js:32]: Whoops, sorry. I never actually tested it myself.

File: integration-testing/election-manager/cypress/tsconfig.json (hunk 15 above)
Comment (eventualbuddha) [reply integration-testing/election-manager/cypress/tsconfig.json:23]: The paths are valid, but why are they here?

File: integration-testing/election-manager/cypress/integration/configuration.ts (hunk 7 above)
Comment (carolinemodic) [reply integration-testing/election-manager/cypress/integration/configuration.ts]: I'll add a comment, the page re-renders twice quickly when it first loads (I haven't dug extensively into why) and I was running into a test flakiness issue where the test would grab the "Definition" button before the rerender happens so when it tries to click it has been detached from the dom. I think that is what is happening in this CI failure: https://app.circleci.com/pipelines/github/votingworks/vxsuite/1555/workflows/2987319e-fcfc-45a5-9a0a-9477d0d66a14/jobs/17622
